### PR TITLE
Configuration plugins

### LINKED ISSUE #56: Plugin configuration registration
With #50 / #51  COBalD just accepts dangling content in YAML configuration. There is no validation whether this actually gets used, and thus no protection against typos. This defeats much of the advantage of using YAML over Python configuration.

I recommend requiring plugins to register which additional configuration sections they expect. This could also directly register a hook to read the configuration.

```python3
setup(
    ...,
    entry_points={
        'cobald.config.yaml_constructors': [
            'MyExtension = mypackage.mymodule:ExtensionClass',
        ],
        'cobald.config.yaml_section': [
            'MySection = mypackage.mymodule:ConfigReader',
        ],
    },
    ...
)
```

See also issue #52.

### PR DESCRIPTION
**Pull request by maxfischer2781**
Adds plugin support for configuration sections. This pull request includes:

* [x] generalisation/cleanup of configuration loading,
* [x] packages may register plugins to parse configuration sections,
* [x] dangling *but unregistered* config sections are an error,
* [x] improved documentation for YAML plugins.

Closes #56. Partially replaces #51 and #50.

### CODE REVIEW

**Comment by maxfischer2781** [comment]
As a side-effect, the cobald pipeline can now be loaded as a plugin to the daemon. See issue #10.

**Comment by maxfischer2781** [comment]
Closes issue #56.

File: src/cobald/daemon/core/config.py [hunk 1]
```diff
@@ -69,6 +90,17 @@ def load(config_path: str):
     yield
 
 
+def load_pipeline(content):
```
Comment (giffels) [inline src/cobald/daemon/core/config.py]: Could add type hints, please?

File: docs/source/custom/package.rst [hunk 2]
```diff
@@ -137,6 +131,74 @@ sequences are used as positional arguments.
...
+    this is used automatically.
+
+Section Plugins
+---------------
+
+Section Plugins allow to accept and digest new configuration sections.
+In addition, the ``cobald`` daemon verify that there are no unexpected
+configuration sections to protect against typos and misconfiguration.
+Extensions are entire top-level sections in the YAML file,
+which are passed to the plugin after parsing and tag evaluation:
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply docs/source/custom/package.rst]: It gets whatever type the section is -- could be a ``dict``, ``list`` or primitive type. I've changed the code snippet to mention what is passed to the plugin.

File: docs/source/changes/51.dangling_config.yaml [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
-  a log message. This allows adding configuration sections for plugins.
+  Packages may define Section Plugins for the COBalD YAML configuration. A plugin
+  receives the requested section of the configuration after parsing/evaluating it.
+  The COBalD daemon verifies that no unknown configuration sections are present
+  in the YAML configuration.
 issues:
 - 50
+- 56
 pull requests:
 - 51
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply docs/source/changes/51.dangling_config.yaml]: I've removed the old PR from the fragment.

File: src/cobald/daemon/config/mapping.py [hunk 4]
```diff
@@ -29,7 +33,7 @@ class Translator(object):
     """
     Translator from a mapping to an initialised object hierarchy
     """
-    def translate_hierarchy(self, structure, *, where='', **construct_kwargs):
+    def translate_hierarchy(self, structure: T, *, where='', **construct_kwargs) -> T:
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply src/cobald/daemon/config/mapping.py]: Done.

File: src/cobald/daemon/core/config.py (hunk 1 above)
Comment (maxfischer2781) [reply src/cobald/daemon/core/config.py]: Done.

**Comment by giffels** [comment]
@mschnepf: Please, add your review in order to proceed with this!

**Pull request merged**